pkg/postgres/gorm/largeobject/large_objects.go

```diff
@@ -0,0 +1,169 @@
+package largeobject
+
+import (
+	"errors"
+	"io"
+
+	"gorm.io/gorm"
+)
+
+// LargeObjects is used to access the large objects API with gorm CRM.
+//
+// This is originally created with similar API with existing github.com/jackc/pgx
+// For more details see: http://www.postgresql.org/docs/current/static/largeobjects.html
+type LargeObjects struct {
+	*gorm.DB
+}
+
+// Mode is the open mode for large object
+type Mode int32
+
+const (
+	// ModeWrite is bitmap for write operation on large object
+	ModeWrite Mode = 0x20000
+	// ModeRead is bitmap for read operation on large object
+	ModeRead Mode = 0x40000
+)
+
+// Create creates a new large object with an unused OID assigned
+func (o *LargeObjects) Create() (oid uint32, err error) {
+	result := o.Raw("SELECT lo_create($1)", 0).Scan(&oid)
+	return oid, result.Error
+}
+
+// Open opens an existing large object with the given mode. ctx will also be used for all operations on the opened large
+// object.
+func (o *LargeObjects) Open(oid uint32, mode Mode) (*LargeObject, error) {
+	var fd int32
+	result := o.Raw("select lo_open($1, $2)", oid, mode).Scan(&fd)
+	if result.Error != nil {
+		return nil, result.Error
+	}
+	return &LargeObject{fd: fd, tx: o.DB}, nil
+}
+
+// Unlink removes a large object from the database.
+func (o *LargeObjects) Unlink(oid uint32) error {
+	var count int32
+	result := o.Raw("select lo_unlink($1)", oid).Scan(&count)
+	if result.Error != nil {
+		return result.Error
+	}
+	if count != 1 {
+		return errors.New("failed to remove large object")
+	}
+	return nil
+}
+
+// Upsert insert a large object with oid. If the large object exists,
+// replace it.
+func (o *LargeObjects) Upsert(oid uint32, r io.Reader) error {
+	obj, err := o.Open(oid, ModeWrite)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = obj.Close()
+	}()
+	_, err = obj.Truncate(0)
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(obj, r)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// Get gets the content of the large object and write it to the writer.
+func (o *LargeObjects) Get(oid uint32, w io.Writer) error {
+	obj, err := o.Open(oid, ModeRead)
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(w, obj)
+	if err != nil {
+		return obj.wrapClose(err)
+	}
+	return obj.wrapClose(err)
+}
+
+// A LargeObject implements the large object interface to Postgres database. It implements these interfaces:
+//
+//	io.Writer
+//	io.Reader
+//	io.Seeker
+//	io.Closer
+type LargeObject struct {
+	tx *gorm.DB
```

Comment:
instead of gorm.DB shouldn't this be gorm.Tx?

Reply:
The same as the LargeObject. gorm.Tx has only functions operation transactions and contexts.